tsl/src/compression/compression.c

```diff
@@ -2239,118 +2236,559 @@ compressed_insert_key_columns(Relation relation)
-	 * inserted, so it cannot be empty.
-	 */
-	Assert(!TTS_EMPTY(slot));
+	List *index_oids;
+	ListCell *lc;
+	ScanKeyData *scankeys = NULL;
+	/* get list of indexes defined on compressed chunk */
+	index_oids = RelationGetIndexList(in_rel);
+	*num_scan_keys = 0;
 
-	Relation out_rel = cis->rel;
+	foreach (lc, index_oids)
+	{
+		Relation index_rel = index_open(lfirst_oid(lc), AccessShareLock);
+		bool matches = false;
 
-	if (!ts_indexing_relation_has_primary_or_unique_index(out_rel))
+		/* Must have at least two attributes. */
+		if (index_rel->rd_index->indnatts < 2)
+		{
+			index_close(index_rel, AccessShareLock);
+			continue;
+		}
+
+		scankeys = palloc0((index_rel->rd_index->indnatts) * sizeof(ScanKeyData));
+
+		for (int i = 0; i < index_rel->rd_index->indnatts; i++)
+		{
+			AttrNumber attnum = AttrOffsetGetAttrNumber(i);
+			char *attname = get_attname(RelationGetRelid(index_rel), attnum, false);
+
+			/* If we are at the last attribute, check its the sequence number attribute.
+			 * This means we found all other attributes on the hypertable and this could be our
+			 * index.
+			 */
+			if (index_rel->rd_index->indnatts - 1 == i)
+			{
+				if (strcmp(attname, COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME) == 0)
+					matches = true;
+				break;
+			}
+
+			if (!ts_array_is_member(settings->fd.segmentby, attname))
+			{
+				break;
+			}
+
+			bool isnull;
+			AttrNumber ht_attno = get_attnum(hypertable_relid, attname);
+			Datum value = slot_getattr(slot, ht_attno, &isnull);
+
+			if (isnull)
+			{
+				ScanKeyEntryInitialize(&scankeys[(*num_scan_keys)++],
+									   SK_ISNULL | SK_SEARCHNULL,
+									   attnum,
+									   InvalidStrategy, /* no strategy */
+									   InvalidOid,		/* no strategy subtype */
+									   InvalidOid,		/* no collation */
+									   InvalidOid,		/* no reg proc for this */
+									   (Datum) 0);		/* constant */
+				continue;
+			}
+
+			Oid atttypid = index_rel->rd_att->attrs[AttrNumberGetAttrOffset(attnum)].atttypid;
+
+			TypeCacheEntry *tce = lookup_type_cache(atttypid, TYPECACHE_BTREE_OPFAMILY);
+			if (!OidIsValid(tce->btree_opf))
+				elog(ERROR, "no btree opfamily for type \"%s\"", format_type_be(atttypid));
+
+			Oid opr =
+				get_opfamily_member(tce->btree_opf, atttypid, atttypid, BTEqualStrategyNumber);
+
+			/*
+			 * Fall back to btree operator input type when it is binary compatible with
+			 * the column type and no operator for column type could be found.
+			 */
+			if (!OidIsValid(opr) && IsBinaryCoercible(atttypid, tce->btree_opintype))
+			{
+				opr = get_opfamily_member(tce->btree_opf,
+										  tce->btree_opintype,
+										  tce->btree_opintype,
+										  BTEqualStrategyNumber);
+			}
+
+			/* No operator could be found so we can't create the scankey. */
+			if (!OidIsValid(opr))
+				continue;
+
+			opr = get_opcode(opr);
+			/* We should never end up here but: no opcode, no optimization */
+			if (!OidIsValid(opr))
```

Comment:
Failing to find the opcode here would be a catalog corruption, so should this be `Ensure`? I'd use a different variable for it as well, the operator oid and opcode are different things and it can be confusing if you miss the above line.